float32 pull_angle
float32 speed_fact